### PR TITLE
Compiler::parseServices: allow prefixing service name in the config

### PR DESCRIPTION
**Pull request by matej21**
ref nette/web-content#377
cc @Majkl578 @greeny

### CODE REVIEW

**Comment by greeny** [comment]
Nice one! :+1:

**Comment by fprochazka** [comment]
Nice!

**Comment by dg** [comment]
What about something like

```
services:
	articles: MyBlog\ArticlesModel(@connection)
	comments: MyBlog\CommentsModel(@connection, @self.articles)
```

**Comment by matej21** [comment]
@dg maybe it looks a bit better, but [this](https://github.com/nette/di/pull/85/files#diff-faebc8d5248ff70a5817645c8b8be6ffR10) won't work. But it is probably useless anyway.

**Comment by dg** [comment]
Hmmm, passing service names to methods seems like antipattern.

**Comment by matej21** [comment]
updated

**Comment by dg** [comment]
Great!

I'm just not quite sure that `self` is the best name for this prefix.

**Comment by matej21** [comment]
I'm not sure either since it is used for self-referencing service.

btw, maybe this could be moved to the ContainerBuilder so you can use `@self.foo` syntax in the CompilerExtension as well (instead of `$this->prefix('@foo')`)

**Comment by greeny** [comment]
> btw, maybe this could be moved to the ContainerBuilder so you can use @self.foo syntax in the CompilerExtension as well (instead of $this->prefix('@foo'))

Good idea :+1:

What about `@this`, `@current` or `@scope`?

**Comment by dg** [comment]
> btw, maybe this could be moved to the ContainerBuilder so you can use @self.foo syntax in the CompilerExtension as well (instead of $this->prefix('@foo'))

I can imagine that it could cause complications, for example when somebody will `addSetup('@self ...')` to service from a different scope.

**Comment by dg** [comment]
`@current`, `@this` and `@self` sound similary. `@scope` sound fine, but has a different meaning in DI http://www.javaranch.com/journal/2008/10/dependency-injection-what-is-scope.html.

File: src/DI/Compiler.php [hunk 1]
```diff
@@ -250,6 +250,9 @@ public static function parseServices(ContainerBuilder $builder, array $config, $
 			} elseif ($namespace) {
 				$name = $namespace . '.' . $name;
 			}
+			if ($namespace) {
```
Comment (dg) [inline src/DI/Compiler.php]: Can you move it bellow next condition?

Comment (matej21) [inline src/DI/Compiler.php]: done

**Comment by dg** [comment]
Thanks!

**Comment by matej21** [comment]
great! now I can finally fix the doc :)

**Pull request merged**